backpack/core/derivatives/batchnorm_nd.py

```diff
@@ -0,0 +1,312 @@
+"""Contains derivatives for BatchNorm."""
+from typing import List, Tuple, Union
+
+from torch import Size, Tensor, einsum
+from torch.nn import BatchNorm1d, BatchNorm2d, BatchNorm3d
+
+from backpack.core.derivatives.basederivatives import BaseParameterDerivatives
+
+
+class BatchNormNdDerivatives(BaseParameterDerivatives):
+    """Derivatives for BatchNorm.
+
+    If training=False: saved statistics are used.
+    If training=True: statistics of current batch are used.
+
+    Index convention:
+    n: batch axis
+    c: category axis
+    {empty}/l/hw/dhw: dimension axis for 0/1/2/3-dimensions
+
+    As a starting point, see these references:
+    https://kevinzakka.github.io/2016/09/14/batch_normalization/
+    https://chrisyeh96.github.io/2017/08/28/deriving-batchnorm-backprop.html
+    """
+
+    def __init__(self, n_dim: int):
+        """Initialization.
+
+        Args:
+            n_dim: number of dimensions
+        """
+        self.n_dim = n_dim
+
+    def _check_parameters(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> None:
+        pass
+
+    def hessian_is_zero(self) -> bool:
+        """Whether hessian is zero.
+
+        Returns:
+            whether hessian is zero
+        """
+        return False
+
+    def hessian_is_diagonal(self) -> bool:
+        """Whether hessian is diagonal.
+
+        Returns:
+            whether hessian is diagonal
+        """
+        return False
+
+    def _jac_mat_prod(
+        self,
+        module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d],
+        g_inp: Tuple[Tensor],
+        g_out: Tuple[Tensor],
+        mat: Tensor,
+    ) -> Tensor:
+        return self._jac_t_mat_prod(module, g_inp, g_out, mat)
+
+    def _get_normalized_input_and_var(
+        self, module: Union[BatchNorm1d, BatchNorm2d, BatchNorm3d]
+    ) -> Tuple[Tensor, Tensor]:
+        _n_axis: int = self._get_n_axis(module)
+        dim: Tuple[int] = {
+            0: (0,),
+            1: (0, 2),
+            2: (0, 2, 3),
+            3: (0, 2, 3, 4),
+        }[_n_axis]
+        input: Tensor = module.input0
+        if module.training:
+            mean: Tensor = input.mean(dim=dim)
+            var: Tensor = input.var(dim=dim, unbiased=False)
+        else:
+            mean: Tensor = module.running_mean
+            var: Tensor = module.running_var
+        mean_expanded: Tensor = {
+            0: mean[None, :],
+            1: mean[None, :, None],
+            2: mean[None, :, None, None],
+            3: mean[None, :, None, None, None],
+        }[_n_axis]
+        var_expanded: Tensor = {
+            0: var[None, :],
+            1: var[None, :, None],
+            2: var[None, :, None, None],
+            3: var[None, :, None, None, None],
+        }[_n_axis]
```

Comment:
Question: Is this to avoid broadcasting?

Reply:
Two reasons:
- Broadcasting doesn't work here (for whatever reason)
- I'm always afraid of broadcasting, because I don't know what happens if different dimensions match in size.

The common underlying issue is my lack of broadcasting knowledge. If you know it better, you can enlighten me.

Separate issue: Probably better do `var=` instead of `var_expanded` (same for mean)?